pagopa/io-dev-api-server
# [IOCOM-445] Polling for PN attachments

⚠️ This PR depends on #298  ⚠️

## Short description
This PR adds polling for PN F24 attachments.

## List of changes proposed in this pull request
F24 polling works by identifying the "/f24/" string inside a requested attachment's relative path.
Polling is handled by a date tuple, in which the first date is the time from when such F24 is ready to be downloaded and the second one is the time after when it is not valid anymore and it has to be polled again.

Let's illustrate the concept with an example. Suppose that we have a F24 identified by `X`. Let 's use `[J,K]` as the tuple for the polling date (`J`) and expiration date (`K`)
- Upon first request for `X`, there is no tuple associated to `X`, so one is generated with `J = (now + attachmentAvailableAfterSeconds)` and `K = (J + attachmentExpiredAfterSeconds)`. Response is `503` with `retry-after: attachmentRetryAfterSeconds` (response header);
- If `X` is requested again with `now < J`, the reply is still `503` with `retry-after: attachmentRetryAfterSeconds`;
- If `X` is requested again with `J < now < K`, then `X` is returned;
- If `X` is requested with `K < now`, then a new tuple `[J1,K1]` is generated (with the same logic as the first request - replacing the previous one) and the response is `503` with `retry-after: attachmentRetryAfterSeconds`.

File Info
- Messages related files have been moved under the `src/features/folder`, structured as follows:
  - `persistence`: messages generation data and in-memory database
  - `routers`: endpoints for messages
  - `services`: business logic for messages endpoints (routers)
  - `types`: not autogenerated types that are related to messages' features
- `src/config`: added polling configuration for PN F24 attachments
- `messagesRouter`: business logic moved to the messagesService file
- `messagesService`: business logic for messages' router. New business logic for messages polling

## How to test
Run the server instance with a PN generation that generates both attachments and F24s. Use the `third-party-messages/:messageId/attachments/:attachmentUrl` to retrieve data. For an attachment, it should always return it. For an F24, the above mechanism should hold true
